Prototype of merge and eager copy

qwwdfsad opened the pull request.

qwwdfsad commented:
cc @yorickhenning

qwwdfsad commented on kotlinx-coroutines-core/jvm/test/ThreadContextMutableCopiesTest.kt:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    class MyMutableElement(
+        private val mutableData: MutableList<String>
```
Here is the trickier use-case when the element "mirrors" the mutable data.

It may be too grotesque to be actually useful. For example, if the data within the element is never mutable, but always mirrors some thread-local or any other storage -- we can get rid of `merge` completely and just invoke `copyForChildCoroutine()` on the resulting `coroutineContext + context`. WDYT?

elizarov commented on kotlinx-coroutines-core/jvm/src/CoroutineContext.kt:
```diff
@@ -10,35 +10,68 @@ import kotlin.coroutines.jvm.internal.CoroutineStackFrame
+private fun CoroutineScope.foldCopies(appendContext: CoroutineContext): CoroutineContext {
+    // Do we have something to copy left-hand side?
+    val hasElementsLeft = coroutineContext.fold(false, hasCopyableElements)
+    val hasElementsRight = appendContext.fold(false, hasCopyableElements)
```
It is not entirely clear why do you `hasElementsRight` here. It looks that even if the right has copyable element (but the left has not), then effect of this code is a simple `coroutineContext + appendContext`, so you can still take the fast path.

qwwdfsad replied:
I'm still exploring whether we should always do copy when there is an element on the right-hand side.

qwwdfsad commented on kotlinx-coroutines-core/jvm/test/ThreadContextMutableCopiesTest.kt:
```diff
@@ -0,0 +1,113 @@
+                threadLocalData.get().add("Y")
+                // Note here, +root overwrites the data
+                launch(Dispatchers.Default + MyMutableElement(mutableListOf("Z"))) {
+                    assertEquals(listOf("X", "Y", "Z"), threadLocalData.get().sorted())
```
>I think the unit tests require sorted() because runBlocking {} is a corner case?

It requires `sorted()` just because I wrote `merge` function this way. I've rewritten it and it no longer requires `sorted()`

>merge() is also a member function. So, if the left-handed and right-handed values were different subclasses of the same supertype that so-happened to use the same Element key, a different merge() function might be called depending on whether or not the left-handed input was present.

I'm not sure that we want to actively promote use-case when different types of elements have the same key. 
If so, then it's the sole responsibility of `merge` function to take care of the polymorphic typing.

>If in most cases, Coroutines were to call leftHanded.merge(rightHanded) but sometimes it were to call rightHanded.merge(EmptyCoroutineContext), the implementation of merge() would have to branch and introspect to figure out how it was called. 


The only guarantee `merge` provides is that this member will be invoked on an element if it is being overwritten by another element with the same key during a new coroutine creation.

qwwdfsad commented on kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt:
```diff
@@ -111,20 +111,29 @@ public interface ThreadContextElement<S> : CoroutineContext.Element {
  *         return TraceContextElement(traceThreadLocal.get()?.copy())
  *     }
+ *
+ *     override fun merge(element: TraceContextElement): TraceContextElement {
```
I've addressed all your points except the last one with `merge` signature receiving two contexts.

[IMO, opinioanted] it looks a bit too overkill and complicates the reasoning about the resulting context when multiple arbitrary copyable elements are present. I'm not sure that allowing to hijack arbitrary context elements in 'merge' function should be the goal as it opens the really wide road to abuse

qwwdfsad commented:
cc @yorickhenning 

I've pushed renewed version of the API:

* All tests you might be interested in are passing now: `withContext` and `runBlocking` included
* `sorted()` is no longer required
* `merge` is as typesafe as it could be (it has one unchecked cast, but it is sound)
* Elements from the right-hand side are properly copied, so accidental sharing is no longer an issue

I believe I addressed most of your concerns, please take a look.

With the current API, it also seems that it supersedes `CoroutineStartInterceptor`: when properly defined, `copyForChildCoroutine()` can be used as `onStart()` and `merge` can be used as `onStart()` when more than one interceptor is present, though it neither was nor is a goal.

Let's see if the overall semantics/approach looks good enough and then decide on our further actions

elizarov commented on kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt:
```diff
@@ -137,6 +147,16 @@ public interface CopyableThreadContextElement<S> : ThreadContextElement<S> {
+     * This function is called on current element, supplied with an element retrieved from child's
+     * coroutine context by the current [key].
+     */
+    public fun merge(overwritingElement: CoroutineContext.Element): CoroutineContext
```
Even though `CopyableThreadContextElement` is currently experimental and breaking changes are expected, it seems polite and fitting to add a default implementation for `merge` that simply returns `overwritingElement`.

elizarov commented on kotlinx-coroutines-core/jvm/src/CoroutineContext.kt:
```diff
@@ -10,35 +10,73 @@ import kotlin.coroutines.jvm.internal.CoroutineStackFrame
+/*
+ * Folds two contexts if there is need to.
+ */
+private fun foldCopies(originalContext: CoroutineContext, appendContext: CoroutineContext, isNewCoroutine: Boolean): CoroutineContext {
```
Effectively, we are trying to provide here an extension point to implementing a mutable (sic) `CoroutineLocal` without actually giving a concrete implementation of it. The reason for not giving a concrete implementation, as I see it, is that there seem to be two different use-cases for mutable coroutine locals:
* Like in your case, you may have an existing system that keeps some "source of truth" in some behind-the-scenes thread-local storage and all you need to do is do synchronize with it. If we just gave you a mutable `CoroutineLocal`, then it would not have helped in writing this synchronization logic.
* We might envision some genuine need to have mutable coroutine locals that are not synchronized to any thread-local storage. 

An interesting observation here is that if we just limit ourselves to those two use-cases, then we may not need two separate `copy` and `merge` methods at the end.

yorickhenning replied:
> We might envision some genuine need to have mutable coroutine locals that are not synchronized to any thread-local storage.

An unsynchronized mutable coroutine local might be sufficient in all cases.

A (classic, not `Copyable`) `ThreadContextElement` could read and write an adjacent thread-unsafe `CoroutineLocal`. With some "clever" typing, a single public `CoroutineContext.Element` type can probably be implemented by a pair of elements?

> An interesting observation here is that if we just limit ourselves to those two use-cases, then we may not need two separate copy and merge methods at the end.

I don't think separate methods are necessary.

> Effectively, we are trying to provide here an extension point to implementing a mutable (sic) CoroutineLocal without actually giving a concrete implementation of it.

It may be safer to give a concrete implementation than a way to supply an implementation. If an extension point is indeed "too grotesque to be actually useful", I might be cautious about entrusting my program's memory consistency to it.

qwwdfsad commented on kotlinx-coroutines-core/jvm/src/ThreadContextElement.kt:
```diff
@@ -111,12 +111,21 @@ public interface ThreadContextElement<S> : CoroutineContext.Element {
  *         return TraceContextElement(traceThreadLocal.get()?.copy())
  *     }
+ *
+ *     override fun merge(element: TraceContextElement): TraceContextElement {
```
>Is that roughly correct?

Yes. Context elements polymorphism is a mine field that require a discipline to follow, otherwise the element may start to interact unpredictably, depending on how `minusKey` and `plus` are invoked on the context.

>That can happen sometimes for reasons I haven't fully understood. 

That's likely to be `Dispatchers.Unconfined` and `CoroutineStart.UNDISPATCHED`

yorickhenning replied:
> Yes. 

OK.

Then I think there are two considerations:

1. Which is the more error-prone API: `copy()` + `merge()` (though it seems `copy()` alone might be OK?), or a delegated `newCoroutineContext()`?
2. Which would Coroutines prefer to support and have to avoid accidentally breaking with future changes?

Either API seems sufficient, and I don't think a third option will soon present itself.

> Context elements polymorphism is a mine field that require a discipline to follow, otherwise the element may start to interact unpredictably, depending on how minusKey and plus are invoked on the context.

Agreed.

> That's likely to be Dispatchers.Unconfined and CoroutineStart.UNDISPATCHED

I had a suspicion. :)
